### PR TITLE
feat: featured, latest, best selling, top rated product sections added to the store page

### PR DESCRIPTION
**Pull request by devAsadNur**
Previously, there was no additional products section to the single store page.

Now, added featured, latest, best selling and top rated product sections to the store page.

Also, added visibility settings to the theme customizer and vendor dashboard settings for the newly added product sections.

Screenshots:
1. Single store page: https://prnt.sc/228s8i1
2. Theme customizer settings from admin dashboard: https://prnt.sc/228riy1
3. Store settings form in vendor dashboard: https://prnt.sc/232785c

### CODE REVIEW

File: includes/Dashboard/Templates/Settings.php [hunk 1]
```diff
@@ -524,6 +524,10 @@ function insert_settings_info() {
                 'phone'                    => sanitize_text_field( $post_data['setting_phone'] ),
                 'show_email'               => sanitize_text_field( $post_data['setting_show_email'] ),
                 'show_more_ptab'           => sanitize_text_field( $post_data['setting_show_more_ptab'] ),
+                'show_featured_products'   => ! empty( $post_data['setting_show_featured_products'] ) ? sanitize_text_field( $post_data['setting_show_featured_products'] ) : 'no',
```
Comment (ManiruzzamanAkash) [inline includes/Dashboard/Templates/Settings.php:527]: Isn't that `isset()` checking is enough for here?
We should use `isset()` and `empty()` in it's proper places based on scenario. Please check for other settings also.

File: templates/settings/store-form.php [hunk 2]
```diff
@@ -182,6 +188,53 @@
...
 
+        <?php
+        $sections = array_filter( $products_block_appearance, function( $value ) {
+            return empty( $value );
+        } );
+
+        if ( count( $sections ) > 0 ) {
+            ?>
+            <div class="dokan-form-group">
+            <label class="dokan-w3 dokan-control-label"><?php esc_html_e( 'Store page product sections', 'dokan-lite' ); ?></label>
```
Comment (ManiruzzamanAkash) [inline templates/settings/store-form.php:199]: this `<label>` should have proper alignment according to it's parent `<div>`.

and same for others.

File: templates/settings/store-form.php [hunk 3]
```diff
@@ -182,6 +188,53 @@
...
+
+            <div class="dokan-w5 dokan-text-left">
+
+            <?php if ( empty( $products_block_appearance['hide_featured_products'] ) ) { ?>
```
Comment (ManiruzzamanAkash) [inline templates/settings/store-form.php:203]: in templating always use alternate if-else statement like this -
```php
<?php if ( empty( $products_block_appearance['hide_featured_products'] ) ) : ?>
    <!-- HTML Code -->
<?php endif; ?>
```

Same is applicable for - `foreach`

Please re-factor this.

File: templates/settings/store-form.php [hunk 4]
```diff
@@ -182,6 +188,53 @@
...
+                <div class="checkbox">
+                    <label>
+                        <input type="checkbox" name="setting_show_featured_products" value="yes"<?php checked( $show_featured_products, 'yes' ); ?>> <?php esc_html_e( 'Show featured producdts section', 'dokan-lite' ); ?>
```
Comment (ManiruzzamanAkash) [inline templates/settings/store-form.php:206]: please add a space after `value="yes"`

and do this for all other places.

File: templates/store.php [hunk 5]
```diff
@@ -30,7 +35,143 @@
...
+                    <p class="dokan-info"><?php esc_html_e( 'No top rated products were found of this vendor!', 'dokan-lite' ); ?></p>
+
+                <?php } ?>
+                </div>
+                <?php
+                }
+
+                if ( ! empty( $extra_product_block ) ) {
+                    ?>
+                    <h2 class="products-list-heading"><?php esc_html_e( 'All Products' ); ?></h2>
```
Comment (ManiruzzamanAkash) [inline templates/store.php:171]: translation is missing.

File: templates/store.php [hunk 6]
```diff
@@ -30,7 +35,143 @@
...
+                <?php
+                }
+
+                // Best selling products section
+                if ( empty( $products_block_appearance['hide_best_selling_products'] ) && ( empty( $store_settings['show_best_sell_products'] ) || 'yes' === $store_settings['show_best_sell_products'] ) ) {
+                    $extra_product_block   = 1;
+                    $best_selling_products = dokan_get_best_selling_products( $items_to_show, $vendor_id );
+                    ?>
+                <div id="dokan-best-selling-products" class="dokan-products-display_section">
+                    <h2 class="products-list-heading"><?php esc_html_e( 'Best Selling Products' ); ?></h2>
```
Comment (ManiruzzamanAkash) [inline templates/store.php:111]: translation missing.

File: templates/store.php [hunk 7]
```diff
@@ -30,7 +35,143 @@
...
+                <?php
+                }
+
+                // Latest products section
+                if ( empty( $products_block_appearance['hide_latest_products'] ) && ( empty( $store_settings['show_latest_products'] ) || 'yes' === $store_settings['show_latest_products'] ) ) {
+                    $extra_product_block = 1;
+                    $latest_products     = dokan_get_latest_products( $items_to_show, $vendor_id );
+                    ?>
+                <div id="dokan-latest-products" class="dokan-products-display_section">
+                    <h2 class="products-list-heading"><?php esc_html_e( 'Latest Products' ); ?></h2>
```
Comment (ManiruzzamanAkash) [inline templates/store.php:79]: translation missing.

File: templates/store.php [hunk 8]
```diff
@@ -30,7 +35,143 @@
...
+                <?php
+                }
+
+                // Top rated products section
+                if ( empty( $products_block_appearance['hide_top_rated_products'] ) && ( empty( $store_settings['show_top_rated_products'] ) || 'yes' === $store_settings['show_top_rated_products'] ) ) {
+                    $extra_product_block = 1;
+                    $top_rated_products  = dokan_get_top_rated_products( $items_to_show, $vendor_id );
+                    ?>
+                <div id="dokan-top-rated-products" class="dokan-products-display_section">
+                    <h2 class="products-list-heading"><?php esc_html_e( 'Top Rated Products' ); ?></h2>
```
Comment (ManiruzzamanAkash) [inline templates/store.php:143]: translation missing.

File: templates/store.php [hunk 9]
```diff
@@ -30,7 +35,143 @@
...
+                <?php
+                dokan_get_template_part( 'store-header' );
+
+                // Featured products section
+                if ( empty( $products_block_appearance['hide_featured_products'] ) && ( empty( $store_settings['show_featured_products'] ) || 'yes' === $store_settings['show_featured_products'] ) ) {
+                    $extra_product_block = 1;
+                    $featured_products   = dokan_get_featured_products( $items_to_show, $vendor_id );
+                    ?>
+                <div id="dokan-featured-products" class="dokan-products-display_section">
+                    <h2 class="products-list-heading"><?php esc_html_e( 'Featured Products' ); ?></h2>
```
Comment (ManiruzzamanAkash) [inline templates/store.php:47]: Translation missing.

File: templates/store.php [hunk 10]
```diff
@@ -8,10 +8,15 @@
...
-$store_info   = $store_user->get_shop_info();
-$map_location = $store_user->get_location();
-$layout       = get_theme_mod( 'store_layout', 'left' );
+$store_user                = dokan()->vendor->get( get_query_var( 'author' ) );
+$store_info                = $store_user->get_shop_info();
+$map_location              = $store_user->get_location();
+$layout                    = get_theme_mod( 'store_layout', 'left' );
+$vendor_id                 = $store_user->id;
+$products_block_appearance = dokan_get_option( 'store_products_section', 'dokan_appearance' );
+$store_settings            = dokan_get_store_info( $vendor_id );
```
Comment (ManiruzzamanAkash) [inline templates/store.php:17]: Please recheck this `$store_settings ` value if we already got this from `$store_info`.
As behind the scene, this called the `get_shop_info()` function too.

File: templates/store.php [hunk 11]
```diff
@@ -30,7 +34,143 @@
...
+                </div>
+                <?php
+                }
+
+                // Latest products section
+                if ( empty( $products_block_appearance['hide_latest_products'] ) && ( empty( $store_info['show_latest_products'] ) || 'yes' === $store_info['show_latest_products'] ) ) {
+                    $extra_product_block = 1;
+                    $latest_products     = dokan_get_latest_products( $items_to_show, $vendor_id );
+                    ?>
+                <div id="dokan-latest-products" class="dokan-products-display_section">
```
Comment (SadikurSadik) [inline templates/store.php:77]: Repeatable code block. Please introduce one function or template for reusability.

**Review by SadikurSadik — CHANGES_REQUESTED** [review]

**Review by ManiruzzamanAkash — APPROVED** [review]

**Review by SadikurSadik — APPROVED** [review]

**Review by nurul-umbhiya — APPROVED** [review]

File: includes/ProductSections/Manager.php [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+     *
+     * @return bool
+     */
+    protected function is_enabled( $section_object ) {
+        $customizer_settings = dokan_get_option( 'product_sections', 'dokan_appearance' );
+        // for customizer settings default value is hide sections
+        if ( $section_object->get_show_in_customizer() &&
+            (
+                ! isset( $customizer_settings[ $section_object->get_section_id() ] ) ||
+                'on' === $customizer_settings[ $section_object->get_section_id() ]
```
Comment (tanmayjay) [inline includes/ProductSections/Manager.php:310]: Is this supposed to be `'on'` or `'off'`?

File: includes/ProductSections/Manager.php [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+     * @param AbstractProductSection $section_object
+     * @param array $store_settings
+     *
+     * @return bool
+     */
+    protected function is_visible( $section_object, $store_settings ) {
+        $customizer_settings = dokan_get_option( 'product_sections', 'dokan_appearance' );
+        // Check if current products section enabled by admin.
+        // for customizer settings default value is hide sections
+        if ( $section_object->get_show_in_customizer() &&
```
Comment (tanmayjay) [inline includes/ProductSections/Manager.php:272]: If the `is_enabled` method gets moved inside the abstract function as mention by the review at `#L304`, we can remove this whole `if` condition by `if ( $section_object->is_enabled() )`. In that case, `$customizer_settings = dokan_get_option( 'product_sections', 'dokan_appearance' );` this line will not be needed.

**Review by tanmayjay — CHANGES_REQUESTED** [review]

File: includes/ProductSections/Manager.php (hunk 12 above)
Comment (nurul-umbhiya) [reply includes/ProductSections/Manager.php:310]: it is supposed to be on, // for customizer settings default value is hide sections

**Comment by Rat01047** [comment]
Issue:
Syntactic error, all the other words are kept uncapitalised except "Latest Products"

Screenshot for reference:
image

**Comment by Rat01047** [comment]
Issue:
The Rating system's color fill of the stars has changed.

Screenshot given for reference: (Develop Branch vs This Branch)
BeFunky-collage

**Pull request merged**